Refactor differentiators

zaqqwerty opened the pull request:
Puts differentiator logic into its own function to expose intermediate circuits to the user.  Resolves #299

MichaelBroughton commented on docs/tutorials/gradients.ipynb:
```diff
@@ -503,19 +519,19 @@
-      ]
+      ],
+      "execution_count": null,
+      "outputs": []
```
Once you get the notebook working, don't forget to restart it and clear the outputs.

zaqqwerty replied:
This is still how it looks after going to "Runtime > Restart runtime" and then doing "Edit > Clear all outputs" before downloading

MichaelBroughton commented on tensorflow_quantum/python/differentiators/differentiator.py:
```diff
@@ -179,6 +179,27 @@ def refresh(self):
+    @abc.abstractmethod
+    def get_intermediate_logic(self, programs, symbol_names, symbol_values,
+                               pauli_sums):
+        """Returns circuits and values which are used by both diff types.
```
Going to need more description here if a user is going to understand how to implement this effectively.

zaqqwerty replied:
Expanded the docstring

MichaelBroughton commented on tensorflow_quantum/python/differentiators/linear_combination.py:
```diff
@@ -96,16 +96,14 @@ def __init__(self, weights, perturbations):
 
-        # these get used a lot
-        n_symbols = tf.gather(tf.shape(symbol_names), 0)
+        See base class for Args.
```
What is being returned here ? The base class just says what the args are but intermediate logic for each differentiation algorithm could be drastically different and should be documented.

zaqqwerty replied:
Changed the API of `get_intermediate_logic`, the base class now describes both `Args` and `Returns`

MichaelBroughton commented on tensorflow_quantum/python/differentiators/parameter_shift.py:
```diff
@@ -102,57 +157,20 @@ def differentiate_analytic(self, programs, symbol_names, symbol_values,
-            tf.expand_dims(flat_shifts, axis=1)
-        ],
-                                       axis=1)
         flat_ops = tf.reshape(
```
This isn't being returned by `get_intermediate_logic` function even though it is essential for the expectation calculation below, we shouldn't force a user to have to think harder than they need to on how to call into the op in their own way should they want to.

zaqqwerty replied:
Refactored

zaqqwerty commented:
@MichaelBroughton I came up with a new API for the `get_intermediate_logic` function which I think provides the flexibility needed to implement SG diffs cleanly.  I implemented it fully for `LinearCombination`, but I wanted to get your thoughts on it before I go ahead with the unit tests and implementation for PS.   Also, I forked off the initial removal of SGD stuff into its own PR #372 so that this one doesn't get too cluttered.

MichaelBroughton commented on tensorflow_quantum/python/differentiators/differentiator.py:
```diff
@@ -179,6 +179,73 @@ def refresh(self):
+                containing the string representation of the operators that will
+                be used on all of the circuits in the expectation calculations.
+
+        Returns:
+            flat_programs: 1-D `tf.Tensor` of strings representing all the
+                circuits required to be run to evaluate the gradient under the
+                derived differentiator.  The length is determined by the
+                specifics of the derived differentiator.
+            flat_symbol_names: 1-D `tf.Tensor` of of strings, containing the
+                name of every symbol used in every circuit in `flat_programs`.
+                The length is determined by the specifics of the derived
+                differentiator.
+            flat_symbol_values: 2-D `tf.Tensor` of DType `tf.float32` containing
+                values to fill in to every parameter in every circuit.
+                The first dimension is the length of `flat_programs` and the
+                second is the length of `flat_symbol_names`.
+            flat_pauli_sums: 2-D `tf.Tensor` of strings representing the
+                operators to measure against each circuit of `flat_programs`.
+                First dimension is the length of `flat_program`, but the
+                second dimension is determined by the specifics of the
+                derived differentiator.
+            flat_mapper: 5-D `tf.Tensor` of DType `tf.float32`.
+                The first dimension is the length of the input `programs`, the
+                second dimension is the length of the second dimension of the
+                input `pauli_sum`, the third dimension is the length of the
+                input `symbol_names`, and the last two dimensions are the same
+                shape as the output `flat_pauli_sums`.  The value at index
+                `ijkmn` is the amount of weight to give the measurement result
+                at `mn` when summing over expectation values to obtain the
+                gradient of the symbol `k` with respect to pauli sum `j` of
+                program `i`.
+        """
```
Given the potentially drastically different "intermediate logic blocks" between differentiators. I think it might make more sense to just have the return signatures defined on a per differentiator basis and leave the parent class as an abstract method. Trying to fit all differentiators under this roof might prove pretty hard.

zaqqwerty replied:
Even though every differentiator can be defined via this interface, I understand your point that it could be a bit hard to reason about.  I think it could make sense to restrict the interface slightly to assume we will use the same gradient procedure for every circuit; then the mapper tensor would be the same for every circuit in the input batch, allowing us to drop an index and reduce the size of the mapper.  I'll think a bit more about how to simplify this interface along that line.

zaqqwerty replied:
Changed to a new, simpler batch interface.  ParameterShift and LinearCombination have been updated to obey the interface.  I still need to write tests for `get_intermediate_logic` confirming shapes and values explicitly.

zaqqwerty replied:
Now added an explicit equation at the end of the docstring to justify why the returns have the shapes specified.

MichaelBroughton reviewed (changes requested):
Looking good, a major change we should adopt here is letting the differentiator define it's own return signature, as we introduce more it will be harder and harder to keep a standardized return signature. An advanced user would also likely know that and expect different forms as well.

zaqqwerty commented:
Given your comment on the interface difficulty, I simplified things a bit to a batch API, where the same transformation is applied to each circuit in the input.  I refactored `LinearCombination` and `ParameterShift` into compliance with this new interface.  I added an explicit equation at the end of the base class docstring to help the user understand the shapes of the returns and how to use them.  Finally, I added annotated tests for these new `get_intermediate_logic` functions.

MichaelBroughton reviewed (changes requested):
The interface change seems reasonable although I personally still think it is still pretty complicated.

A more major problem that I have with this re-implementation is that we have moved from making just ONE call to the provided expectation op to many seperate calls to the given expectation function inside of a `map_fn`. Looking at the docs page for `map_fn` we see: 

> map_fn unstacks elems on axis 0 to obtain a sequence of elements; calls fn to transform each element; and then stacks the transformed values back together.

In our case, this will drastically degrade performance both in simulation and on a real QPU. Simulation with have to allocate / simulate and deallocate memory for each individual circuit which is very wasteful. If the circuits are small it will also be unable to parallelize very much ( if at all ). When running on QPU the performance will also be worse. If we are unable to see the ENTIRE batch of circuits we need torun , we will never be able totake full advantage of more efficient engine methods like `run_batch` (from Cirq) or `run_sweep` (from Cirq).

If we want to keep this structure I think we should definitely fix this issue inside of the `differentiate_***` methods too.

zaqqwerty commented:
Sure, I can just stack the circuits first instead

zaqqwerty commented:
Reshaped things in the default diff logic so that expectation op is only called once per differentiation again

zaqqwerty commented:
Had to double the number of samples in the new gradient tests to prevent flaking

MichaelBroughton reviewed (commented):
I'm sorry that this has turned into such a long discussion. Perhaps it is a good indicator that we might need a design document....

Now I'm wondering how get_intermediate_logic might behave with something like the SGDifferentiator where the number of samples taken matters a great deal. The `get_intermediate_logic` method does not take num_samples as an argument which might be something we want to include ?

Would you be open to discussing this more in tomorrow meeting ?

zaqqwerty commented:
Sure, let's chat about it tomorrow, I want to make sure it works for everyone's use cases.  I spoke with Jae about this new interface, he was of the opinion it will ease the re-implementation of the SGDifferentiators (those also used just a simple tiling of the input num_samples, but we can discuss including it as an argument to `get_intermediate_logic`)

zaqqwerty commented:
I thought more about your comment on `num_samples`.  I was able to include it as part of the input and output of `get_intermediate_logic`, and this actually allowed the definitions of `differentiate_analytic` and `differentiate_sampled` to be pulled all the way to the base class only, greatly reducing code duplication.

MichaelBroughton reviewed (changes requested):
Looking at the code I remember how tricky it was implementing these things the first time. Would we be open to doing this changeover in stages:

1. Add the interface and functionality of `get_intermediate_logic`. (This PR)
2. Move over all of the `differentiate_analytic` and `differentiate_sampled` to use them (another PR) ?

zaqqwerty commented on tensorflow_quantum/python/differentiators/differentiator.py:
```diff
@@ -180,14 +180,111 @@ def refresh(self):
+            As an explicit equation, the return values are defined in terms of
+            the argument values such that:
+
+            d(<programs[i](symbol_values[i])|
+                pauli_sums[i, j]|programs[i](symbol_values[i])>)
+            / d(symbol_names[k]) = sum_mn (batch_mapper[i, j, k, m, n] *
+                <batch_programs[i][m](batch_symbol_values[i][m])|
+                    batch_pauli_sums[i, m, n]
+                        |batch_programs[i][m](batch_symbol_values[i][m])>)
```
Here `d` is the differentiation symbol

MichaelBroughton replied:
Might want to clear that up so it's easy to understand in the doc renderings.

zaqqwerty commented on tensorflow_quantum/python/differentiators/differentiator.py:
```diff
@@ -180,14 +180,111 @@ def refresh(self):
     @abc.abstractmethod
+    def get_intermediate_logic(self, programs, symbol_names, symbol_values,
+                               pauli_sums, num_samples):
+        """Returns components that can be used to build the gradient.
```
Slightly too long for first line only, added shortened version

zaqqwerty commented on tensorflow_quantum/python/differentiators/differentiator.py:
```diff
@@ -180,14 +180,111 @@ def refresh(self):
+                               pauli_sums, num_samples):
+        """Returns components that can be used to build the gradient.
+
+        In order to compute gradients on quantum computers, one must
+        decompose gradient calculations into circuits to run and measurements
+        to take.  Then, some linear combination of the measurement results
+        can be used to recover estimates of circuit parameter derivatives. This
+        function is intended to encapsulate the decomposition and recombination
+        steps, so that the same logic can be used for both analytic and sample
+        based expectations.  Additionally, centralizing this logic allows
+        advanced users to access the decompositions directly.
+
+        This function defines a batch interface for decomposing circuits
+        into their derivative components: it is assumed that each circuit in
+        the input batch will have the same gradient procedure applied to it.
+        Thus the outer dimension of all returned components is the same as the
+        input `programs` batch dimension.
+
+        For each entry in the batch dimension of the returned values,
+        inheriting differentiators are free to generate new circuits, symbols,
+        and measurement ops as needed. To combine these outputs into the
+        required derivatives, the last return value is a map specifying the
+        linear combination of batched expectation values to use to calculate the
+        derivative of each input parameter.
```
Added a `NOTE:` paragraph pointing this out

zaqqwerty commented on tensorflow_quantum/python/differentiators/differentiator.py:
```diff
@@ -180,14 +180,111 @@ def refresh(self):
+            pauli_sums: `tf.Tensor` of strings with shape [batch_size, n_ops]
+                containing the string representation of the operators that will
+                be used on all of the circuits in the expectation calculations.
+            num_samples: `tf.Tensor` of positive integers representing the
```
Added the shape

zaqqwerty commented on tensorflow_quantum/python/differentiators/differentiator.py:
```diff
@@ -180,14 +180,111 @@ def refresh(self):
+            pauli_sums: `tf.Tensor` of strings with shape [batch_size, n_ops]
+                containing the string representation of the operators that will
+                be used on all of the circuits in the expectation calculations.
+            num_samples: `tf.Tensor` of positive integers representing the
```
Makes sense to make this optional; do you know of any best practices around optional arguments to functions with `tf.function`?  My experience is these can get finicky

MichaelBroughton replied:
I'm pretty sure tf.functions don't support the intuitive python style default arguments, but an easy workaround is to just have a null value tensor that indicates the `get_intermeidate_logic` function won't process the samples tensor. For example you could pass in tf.constant(-2) with shape = `()` to indicate you don't need samples.

zaqqwerty commented:
Hi @MichaelBroughton, I addressed your comments on the docstring.  I also broke this PR in two as you suggested; this one just adds the `get_intermediate_logic` interface, implementations, and tests; #379 deals with removing the old `differentiate_analytic` and `differentiate_sampled` code.

MichaelBroughton reviewed (changes requested):
Tested this locally on my machine and made a worrying discovery: In a relatively moderate setting of a batch of 64 circuits each with 25 symbols , 10 ops and a central-difference differentiatior, just calling `get_intermediate_logic` consumed a huge amount of memory. I worked out the numbers and I think the `batch_mapper` is the culprit with shape [64, 10, 25, 50, 50] =>  40 million entries (the last two might be 25's ?) which is far too many. 

I think we need to reconsider this component. Is there any way we could shrink down this tensor and maintain fully intuitive generality ? Maybe this makes a case for moving to each differentiator having it's own unique return signature on `get_intermediate_logic` ?

zaqqwerty commented:
@MichaelBroughton I think you're right that the discussion would be aided by a design doc.  I've started writing one (sent you a DM)

zaqqwerty commented:
After some more thought, since this PR just adds the functionality of intermediate logic and doesn't remove the way it functions elsewhere, could we put this in as-is (after the docstring and optional `num_samples` arg updates), then open an issue and do the careful design doc before swapping this logic into `differentiate_analytic` and `differentiate_sampled`?

To follow up on the memory issue, in your example, `batch_mapper` would actually have shape [64, 10, 25, 51, 10] which implies 8,160,000 entries.  This is about a factor of 100 larger than the existing intermediate steps of LinearCombination [here](https://github.com/tensorflow/quantum/blob/0712199eed1ce77a3b736d886c757ea48d41987f/tensorflow_quantum/python/differentiators/linear_combination.py#L152), `all_perturbations`, which for your example would have shape `[n_symbols, n_perturbations, n_programs, n_symbols]` ==> [25, 2, 64, 25] which is 80,000 entries.  If this is too large, one way that comes to mind to reduce the memory footprint might be to switch batch mapper to a [sparse tensors](https://www.tensorflow.org/api_docs/python/tf/sparse/SparseTensor), since most of the entries in the batch mapper (and in `all_perturbations`) in this case are zero.

MichaelBroughton commented:
I don't think taking on a factor 100x memory growth and then moving to make `differentiate_analytic` and `differentiate_sampled` depend on them would be a wise move here. Let's take a look at the design document and then circle back once we've found some solutions that work on all fronts.

zaqqwerty commented:
Completed design doc has a slightly different proposed interface, closing this one
